weave/trace_server_bindings/caching_middleware_trace_server.py

```diff
@@ -0,0 +1,421 @@
+from __future__ import annotations
+
+import logging
+from collections.abc import Iterator
+from typing import Any, Callable, TypedDict, TypeVar
+
+import diskcache
+
+from weave.trace.refs import ObjectRef, parse_uri
+from weave.trace.settings import (
+    server_cache_dir,
+    server_cache_size_limit,
+    use_server_cache,
+)
+from weave.trace_server import trace_server_interface as tsi
+
+logger = logging.getLogger(__name__)
+
+TReq = TypeVar("TReq", bound=tsi.BaseModel)
+TRes = TypeVar("TRes", bound=tsi.BaseModel)
+
+
+class CacheRecorder(TypedDict):
+    hits: int
+    misses: int
+    errors: int
+    skips: int
+
+
+def digest_is_cacheable(digest: str) -> bool:
+    """
+    Check if a digest is cachable.
+
+    Examples:
+    - v1 -> True, 1
+    - oioZ7zgsCq4K7tfFQZRubx3ZGPXmFyaeoeWHHd8KUl8 -> False, -1
+    """
+    # If it looks like a version, it is not cachable
+    if digest.startswith("v"):
+        try:
+            int(digest[1:])
+            return False  # noqa: TRY300
+        except ValueError:
+            return True
+    elif digest == "latest":
+        return False
+
+    return True
+
+
+class CachingMiddlewareTraceServer(tsi.TraceServerInterface):
+    """A middleware trace server that provides caching functionality.
+
+    This server wraps another trace server and caches responses to improve performance.
+    It uses diskcache to store responses on disk and implements caching for read-only
+    operations like obj_read, table_query, etc.
+
+    Attributes:
+        _next_trace_server: The underlying trace server being wrapped
+        _cache: The disk cache instance used to store responses
+        _cache_recorder: Metrics tracking cache hits, misses, errors and skips
+    """
+
+    _next_trace_server: tsi.TraceServerInterface
+    _cache_prefix: str
+
+    def __init__(
+        self,
+        next_trace_server: tsi.TraceServerInterface,
+        cache_dir: str | None = None,
```

Comment:
How big does the cache get in your testing?  Do we need to provide a cleanup function for users?

Reply:
I am trusting the implementation here